### PR TITLE
Implement 2D instance shader parameters

### PR DESCRIPTION
**Pull request by huwpascoe**
From [KoBeWi](https://github.com/godotengine/godot/pull/77594) to [EIREXE](https://github.com/godotengine/godot/pull/87966) to here. Issue https://github.com/godotengine/godot/issues/62943.

image
[instvartest.zip](https://github.com/user-attachments/files/17007393/instvartest.zip)

Now feature complete with the exception of compatibility renderer.

### CODE REVIEW

**Comment by KoBeWi** [comment]
Parameter list doesn't work properly. It's not updated right when modifying shaders and some operations may replace it with nulls:

https://github.com/user-attachments/assets/fdbcc7b9-83cd-4be8-bcd6-5d5162a8b6fd

**Comment by huwpascoe** [comment]
Fixed the instances not being freed.

I spent 3 hours trying to figure out why the parameter list is broken and found no solution. Essentially the MaterialStorage fails to update correctly when the shader is changed. The uniforms default values for instance variables are initialized to nil. A workaround is to reload the scene/project.

So the parameter list failing to refresh seems to be an unrelated issue.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 99a7a9ccd60fbe4030e067b3c36d54b67737446d), it works as expected in Forward+ and Mobile.

When running in Compatibility, you get a shader compilation error like in 3D (as expected):

```
SHADER ERROR: Uniform instances are not supported in gl_compatibility shaders.
```

image

**Testing project:** [test_pr_97058.zip](https://github.com/user-attachments/files/17021089/test_pr_97058.zip)

PS: What is this checkbox for? It's present for both 3D and 3D per-instance uniforms. I can't uncheck it (nothing happens when I click it), regardless of whether the value is modified from its default or not.

image

File: doc/classes/CanvasItem.xml
Comment (Calinou) [inline doc/classes/CanvasItem.xml:575]: Fix the note to be relevant to the current state. Per-instance uniforms are still unavailable for fog and sky shaders, but this isn't indicated by the docs yet:

```suggestion
				[b]Note:[/b] Per-instance shader uniforms are currently only available for [code]canvas_item[/code] and [code]spatial[/code] shaders.
```

**Comment by huwpascoe** [comment]
> What is this checkbox for?

Don't know, I think it was like that in 4.2 as well.

**Comment by huwpascoe** [comment]
I'll need guidance on how to squash this thing so that original authors are credited etc. Or a maintainer can do it if they're feeling generous, I don't mind.

File: servers/rendering/renderer_canvas_cull.cpp
Comment (AThousandShips) [inline servers/rendering/renderer_canvas_cull.cpp:1862]:
```suggestion
void RendererCanvasCull::_update_instance_shader_uniforms_from_material(HashMap<StringName, Item::InstanceShaderParameter> &p_is_params, const HashMap<StringName, Item::InstanceShaderParameter> &p_existing_is_params, RID p_material) {
```

**Comment by KoBeWi** [comment]
To update property list, you need to use `notify_property_list_changed`. Here's how it's done in 3D:
https://github.com/godotengine/godot/blob/02b16d2f544e323b7b7f57e6e992b0b8e5d8b954/scene/3d/visual_instance_3d.cpp#L244-L246
However modifying shader does not add the CanvasItem to `_instance_update_list`, I think it also has to be done manually.

Also if you want the instance parameters to have revert button, you need to implement `_property_can_revert()` (return true if instance uniform exists) and `_property_get_revert()` (return default value of uniform).

**Comment by huwpascoe** [comment]
Redid everything from the ground up, refactored all the common code into a class. Everything should work identically to the 3D instances now.

File: servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl:55]: I haven't taken a look at the whole thing, but just a reminder that this needs to be removed before this PR is finished. We need to avoid bloating the instance struct at all costs

Comment (huwpascoe) [reply servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl:55]: Once https://github.com/godotengine/godot/pull/97340 is merged then I'll have the space for it I reckon.

Comment (clayjohn) [reply servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl:55]: Why is that? https://github.com/godotengine/godot/pull/97340 clears up a spot in the instance buffer, but uses that spot for its own needs

Comment (huwpascoe) [reply servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl:55]: ...oh. I need to find some space then.

File: doc/classes/CanvasItem.xml
Comment (KoBeWi) [inline doc/classes/CanvasItem.xml:576]: The corresponding note needs to be updated in GeometryInstance3D.

File: doc/classes/RenderingServer.xml
Comment (KoBeWi) [inline doc/classes/RenderingServer.xml:445]: Literally every property is case-sensitive. This note is redundant IMO (I know it's copied).

File: doc/classes/RenderingServer.xml
Comment (KoBeWi) [inline doc/classes/RenderingServer.xml:460]: It's not equivalent.

**Comment by Chaosus** [comment]
Please squash the commits as it's required by our pipeline (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html).

**Comment by huwpascoe** [comment]
> Please squash the commits

Do I need to do anything special to keep the other contributor's names or do I just squash?

**Comment by clayjohn** [comment]
> > Please squash the commits
> 
> Do I need to do anything special to keep the other contributor's names or do I just squash?

You will need to have the "Co-Authored by:" lines for both of them (one is already there). 

I suggest waiting to squash until you are finished with the PR and it is ready to merge.

File: servers/rendering/renderer_canvas_cull.h
Comment (AThousandShips) [inline servers/rendering/renderer_canvas_cull.h:40]:
```suggestion
	static void _dependency_changed(Dependency::DependencyChangedNotification p_notification, DependencyTracker *p_tracker);
```

File: servers/rendering/renderer_canvas_cull.h
Comment (AThousandShips) [inline servers/rendering/renderer_canvas_cull.h:41]:
```suggestion
	static void _dependency_deleted(const RID &p_dependency, DependencyTracker *p_tracker);
```

File: servers/rendering/renderer_canvas_cull.cpp
Comment (AThousandShips) [inline servers/rendering/renderer_canvas_cull.cpp:49]:
```suggestion
void RendererCanvasCull::_dependency_changed(Dependency::DependencyChangedNotification p_notification, DependencyTracker *p_tracker) {
```

File: servers/rendering/renderer_canvas_cull.cpp
Comment (AThousandShips) [inline servers/rendering/renderer_canvas_cull.cpp:61]:
```suggestion
void RendererCanvasCull::_dependency_deleted(const RID &p_dependency, DependencyTracker *p_tracker) {
```

File: servers/rendering/renderer_scene_cull.h
Comment (AThousandShips) [inline servers/rendering/renderer_scene_cull.h:43]:
```suggestion
#include "servers/rendering/instance_uniforms.h"
```
Full path

**Review by AThousandShips — APPROVED** [review]

File: servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl
Comment (stuartcarnie) [reply servers/rendering/renderer_rd/shaders/canvas_uniforms_inc.glsl:55]: Here is the space that you can use once #97340 merges:

https://github.com/godotengine/godot/pull/97340/files#diff-a960cbf133befa41681f0d917a05f34b59265036361c79b2c32ace27e0f3a40fR385

**Comment by bunnybreaker** [comment]
Any idea on if/when this will get added to the main branch?

**Comment by huwpascoe** [comment]
> Any idea on if/when this will get added to the main branch?

Depends on #97340. Soon as it's merged, I'll submit the final revision and then it's in the hands of the maintainers.

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:2431]: This should be passed in as an argument to `new_instance_data()` so that it is always initialized. Right now this will only impact certain canvas_items (like Sprite2Ds)

File: servers/rendering/instance_uniforms.h
Comment (clayjohn) [inline servers/rendering/instance_uniforms.h:46]:
```suggestion
	bool materials_finish(RID p_self);
```
Style nitpick, we shouldn't include [[nodiscard]] unless it is needed. Otherwise you are making the code harder to read for no benefit

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp
Comment (huwpascoe) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:2431]: Should it still be set here?

**Comment by huwpascoe** [comment]
done!

**Comment by huwpascoe** [comment]
It's finished. But they want compat renderer support added.

Someone else can do that, I'm tired.

**Comment by Delsin-Yu** [comment]
> It's finished. But they want compat renderer support added.
> 
> Someone else can do that, I'm tired.

Is there any chance can we merge this and leave the compact renderer support for future contributors?

**Comment by KoBeWi** [comment]
It will at least need a warning that Compat is not supported.

**Comment by huwpascoe** [comment]
https://github.com/godotengine/godot/pull/99230 is now merged, 2d shader parameters will be in 4.4.

**Pull request closed without merging**
